README.md
# Todo Scanner

```
 /$$$$$$$$        /$$$$$$$  /$$$$$$$         /$$$$$$                      /$$   /$$                    
|__  $$__/       | $$__  $$| $$__  $$       /$$__  $$                    | $$$ | $$                    
   | $$  /$$$$$$ | $$  \ $$| $$  \ $$      | $$  \__/  /$$$$$$$  /$$$$$$ | $$$$| $$  /$$$$$$   /$$$$$$ 
   | $$ /$$__  $$| $$  | $$| $$  | $$      |  $$$$$$  /$$_____/ |____  $$| $$ $$ $$ /$$__  $$ /$$__  $$
   | $$| $$  \ $$| $$  | $$| $$  | $$       \____  $$| $$        /$$$$$$$| $$  $$$$| $$$$$$$$| $$  \__/
   | $$| $$  | $$| $$  | $$| $$  | $$       /$$  \ $$| $$       /$$__  $$| $$\  $$$| $$_____/| $$      
   | $$|  $$$$$$/| $$$$$$$/| $$$$$$$/      |  $$$$$$/|  $$$$$$$|  $$$$$$$| $$ \  $$|  $$$$$$$| $$      
   |__/ \______/ |_______/ |_______/        \______/  \_______/ \_______/|__/  \__/ \_______/|__/      
                                                                                                       
```

Todo Scanner is a command-line tool written in Haskell for searching for specific labels (e.g., "TODO," "FIXME," "XXX") in files within a specified directory with particular file extensions.

## Requirements

To use this tool, you need to have the following installed:

- [GHC (Glasgow Haskell Compiler)](https://www.haskell.org/ghc/)
- [cabal](https://www.haskell.org/cabal/)

## Installation

1. Download or clone this repository.
2. Execute the following commands to build and install the program:

```bash
   cd todo-scanner-directory
   cabal update
   cabal build
   cabal install
```

## Usage

```bash
todo-scanner [--dir DIRECTORY] [--label LABEL] [--exts EXTENSION]
```

Parameters:

1. `--dir DIRECTORY`: The directory to scan for files. (can be specified multiple times)
1. `--label LABEL`: Optional. A label to search for (can be specified multiple times).
1. `--exts EXTENSION`: Optional. File extensions to scan for (can be specified multiple times).

Examples:

* Search for TODOs in all files in the "my_project" directory:

```bash
todo-scanner --dir my_project
```

* Search for TODOs and FIXMEs in files with "js" and "php" extensions:

```bash
todo-scanner --dir my_project --label TODO --label FIXME --exts js --exts php
```

## How It Works

1. The program obtains input parameters from the command line using the "optparse-applicative" library.
1. It recursively searches for files in the specified directory with the specified file extensions.
1. For each found file, its content is read.
1. The program analyzes the content of each file, searching for matches with the specified labels.
1. Found matches are printed to the console, along with information about the line and file where they were found.
1. All found matches are highlighted in green, and other lines are highlighted in red.
1. After the search is complete, the program reports the total number of found matches.

## Example Output

If matches (e.g., "TODO") are found in files, the program's output may look like this:

```plaintext
------ --------------------------------------------------
Line: my_project/src/file.js
1     // TODO: Fix this issue
3     // TODO: Refactor this code
------ --------------------------------------------------
Found 2 TODOs

```

If no matches are found, the program reports:

```plaintext
        Found 0 TODOs
```

## Docker

`docker run -v $(pwd):/code --rm -it --entrypoint "/srv/todo-scaner" jtrw/todo-sacenr --dir=/code`

## Gitlab-Ci

```yaml
todo:
  stage: static-analizator
  image:
    name: jtrw/todo-scaner:latest
    entrypoint: [""] # disable entrypoint
  script:
    - /srv/todo-scaner --dir=./src
```



## Notes

* The program uses colored highlighting to make the output more readable. It may not work on all terminals.
* You can customize the default labels and file extensions by using your values.

This documentation provides an overview of the features and usage of Todo Scanner. For more information about capabilities and configuration, it is recommended to review the source code and Haskell documentation.

This tool is designed to make it easier for you to locate "TODO" labels in your project and help you focus on important tasks.

CHANGELOG.md
# Revision history for todo-scaner

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


main.hs
import Data.List
import System.Directory
import System.FilePath
import Control.Monad
import System.IO
import System.Environment
import System.Exit
import Data.Char
import Options.Applicative
import Data.Binary.Put (PutM(Put))

data Options = Options
  { 
    dir :: FilePath,
    label :: [String],
    ext :: [String]
  }

optionsParser :: Parser Options
optionsParser = Options
    <$> strOption
        ( long "dir"
        <> metavar "DIRECTORY"
        <> help "Directory to scan" )
    <*> many (strOption
        ( long "label"
            <> metavar "LABEL"
            <> help "Label to scan for"
        ))
    <*> many (strOption
    ( long "exts"
        <> metavar "EXTENSION"
        <> help "File extension to scan for"
    ))

main :: IO ()
main = do
    options <- execParser opts
    let dirToScan = dir options
    putStrLn $ colorBlue ("Scanning directory: " ++ dirToScan)

    let labels = if null (label options) then getDefaultLabels else label options
    let extensions = if null (ext options) then getDedaultExtensions else ext options

    files <- getRecursiveContents dirToScan extensions

    total <- foldM (\acc file -> do
        count <- checkFile file labels
        return (acc + count)) 0 files

    let msg = "Found " ++ show total ++ " TODOs"
    let lengthSpaces = length msg + 2

    if total == 0 then do
        putStrLn $ spacesWithGreenBackground lengthSpaces
        putStrLn $ textWithGreenBackground msg
        putStrLn $ spacesWithGreenBackground lengthSpaces
        exitSuccess
    else do
        putStrLn $ spacesWithRedBackground lengthSpaces
        putStrLn $ textWithRedBackground msg
        putStrLn $ spacesWithRedBackground lengthSpaces
        exitFailure
    
    exitSuccess
    where
        opts = info (helper <*> optionsParser)
          ( fullDesc
         <> progDesc "Scan directory for TODOs"
         <> header "todo - a simple TODO scanner" )

getRecursiveContents :: FilePath -> [String] -> IO [FilePath]
getRecursiveContents topdir extensions = do
    -- is directory exist
    isDirectoryExist <- doesDirectoryExist topdir
    if not isDirectoryExist
        then do 
            putStrLn $ colorRed ("Directory " ++ topdir ++ " does not exist")
            exitFailure
        else do
            names <- getDirectoryContents topdir
            let properNames = filter (`notElem` [".", ".."]) names
            paths <- forM properNames $ \name -> do
                let path = topdir </> name
                isDirectory <- doesDirectoryExist path

                if isDirectory
                    then getRecursiveContents path extensions
                    else if any (`isSuffixOf` path) extensions
                        then return [path]
                        else return []
            return (concat paths)
    
checkFile :: FilePath -> [String] -> IO Int
checkFile file labels = do
    content <- readFile file
    let linesOfFile = lines content
    let fileLinesWithIndex = zip [1..] linesOfFile
    let fileLinesWithIndexFiltered = filter (\(i, line) -> any (`isInfixOf` line) labels) fileLinesWithIndex
    let fileLinesWithIndexFilteredMapped = map (\(i, line) -> (i, removeBeforeLabels line labels)) fileLinesWithIndexFiltered
    let fileLinesWithIndexFilteredMappedTrim = map (\(i, line) -> (i, dropWhile isSpace line)) fileLinesWithIndexFilteredMapped
    
    if null fileLinesWithIndexFilteredMappedTrim
        then return 0
        else do
            putStrLn "------ --------------------------------------------------"    
            let msg = "Line: " ++ file
            putStrLn $ colorGreen msg
            mapM_ (putStrLn . (\(i, line) -> show i ++ "     " ++ line)) fileLinesWithIndexFilteredMappedTrim
            putStrLn "------ --------------------------------------------------"
            let count = length fileLinesWithIndexFilteredMappedTrim
            return count
            
getDefaultLabels :: [String]
getDefaultLabels = ["TODO", "FIXME", "XXX"]

getDedaultExtensions :: [String]
getDedaultExtensions = ["php", "js", "xml", "yml", "yaml"]

colorGreen :: String -> String
colorGreen input = "\x1b[32m" ++ input ++ "\x1b[0m"

colorRed :: String -> String
colorRed input = "\x1b[31m" ++ input ++ "\x1b[0m"

textWithBackground :: String -> String -> String
textWithBackground color input = "\x1b[" ++ color ++ "m" ++ input ++ "\x1b[0m"

textWithRedBackroundAndBorder :: String -> String
textWithRedBackroundAndBorder input = textWithBackground "41;1;37" input

colorBlue :: String -> String
colorBlue input = "\x1b[34m" ++ input ++ "\x1b[0m"

textWithBlueBackground :: String -> String
textWithBlueBackground input = textWithBackground "44;1;37" input

textWithRedBackground :: String -> String
textWithRedBackground input = textWithBackground "41;1;37" (" " ++ input ++ " ")

textWithGreenBackground :: String -> String
textWithGreenBackground input = textWithBackground "42;1;37" (" " ++ input ++ " ")

spacesWithRedBackground :: Int -> String
spacesWithRedBackground count = textWithBackground "41;1;37" (replicate count ' ')

spacesWithGreenBackground :: Int -> String
spacesWithGreenBackground count = textWithBackground "42;1;37" (replicate count ' ')

removeBeforeLabel :: String -> String -> String
removeBeforeLabel input label = case dropWhile (not . isPrefixOf label) (tails input) of
    (rest:_) -> rest
    _ -> input

removeBeforeLabels :: String -> [String] -> String
removeBeforeLabels input labels = foldl (\acc label -> removeBeforeLabel acc label) input labels

app/Main.hs
import Data.List
import System.Directory
import System.FilePath
import Control.Monad
import System.Exit
import Data.Char
import Options.Applicative
import Data.Bifunctor (second)

data Options = Options
  {
    dir :: [FilePath],
    label :: [String],
    ext :: [String]
  }

optionsParser :: Parser Options
optionsParser = Options
    <$> many (strOption
        ( long "dir"
        <> metavar "DIRECTORY"
        <> help "Directory to scan" ))
    <*> many (strOption
        ( long "label"
            <> metavar "LABEL"
            <> help "Label to scan for"
        ))
    <*> many (strOption
    ( long "exts"
        <> metavar "EXTENSION"
        <> help "File extension to scan for"
    ))

main :: IO ()
main = do
    options <- execParser opts

    let  dirsToScan = dir options

    if null dirsToScan
        then do
            putStrLn $ colorRed "No directory to scan. Use --dir option"
            exitFailure
        else do
            return ()

    mapM_ (putStrLn . (\ currentDir -> colorBlue ("Scanning directory: " ++ currentDir))) dirsToScan

    let labels = if null (label options) then getDefaultLabels else label options
    let extensions = if null (ext options) then getDedaultExtensions else ext options

    files <- foldM (\acc currentDir -> do
        files <- getRecursiveContents currentDir extensions
        return (acc ++ files)) [] dirsToScan

    total <- foldM (\acc file -> do
        count <- checkFile file labels
        return (acc + count)) 0 files

    let msg = "Found " ++ show total ++ " TODOs"
    let lengthSpaces = length msg + 2

    _ <- if total == 0 then do
        putStrLn $ spacesWithGreenBackground lengthSpaces
        putStrLn $ textWithGreenBackground msg
        putStrLn $ spacesWithGreenBackground lengthSpaces
        exitSuccess
    else do
        putStrLn $ spacesWithRedBackground lengthSpaces
        putStrLn $ textWithRedBackground msg
        putStrLn $ spacesWithRedBackground lengthSpaces
        exitFailure
    exitSuccess
    where
        opts = info (helper <*> optionsParser)
          ( fullDesc
         <> progDesc "Scan directory for TODOs"
         <> header "todo - a simple TODO scanner" )

getRecursiveContents :: FilePath -> [String] -> IO [FilePath]
getRecursiveContents topdir extensions = do
    isDirectoryExist <- doesDirectoryExist topdir
    if not isDirectoryExist
        then do
            putStrLn $ colorRed ("Directory " ++ topdir ++ " does not exist")
            exitFailure
        else do
            names <- getDirectoryContents topdir
            let properNames = filter (`notElem` [".", ".."]) names
            paths <- forM properNames $ \name -> do
                let path = topdir </> name
                isDirectory <- doesDirectoryExist path

                if isDirectory
                    then getRecursiveContents path extensions
                    else if any (`isSuffixOf` path) extensions
                        then return [path]
                        else return []
            return (concat paths)

checkFile :: FilePath -> [String] -> IO Int
checkFile file labels = do
    content <- readFile file
    let linesOfFile = lines content
    let fileLinesWithIndex = zip [1..] linesOfFile
    let fileLinesWithIndexFiltered = filter (\(_, line) -> any (`isInfixOf` line) labels) fileLinesWithIndex
    let fileLinesWithIndexFilteredMapped = map (\(i, line) -> (i, removeBeforeLabels line labels)) fileLinesWithIndexFiltered
    let fileLinesWithIndexFilteredMappedTrim = map (Data.Bifunctor.second (dropWhile isSpace)) fileLinesWithIndexFilteredMapped

    if null fileLinesWithIndexFilteredMappedTrim
        then return 0
        else do
            putStrLn "------ --------------------------------------------------"
            let msg = "Line: " ++ file
            putStrLn $ colorGreen msg
            mapM_ (putStrLn . (\(i, line) -> show (i :: Integer) ++ "     " ++ line)) fileLinesWithIndexFilteredMappedTrim
            putStrLn "------ --------------------------------------------------"
            let count = length fileLinesWithIndexFilteredMappedTrim
            return count

getDefaultLabels :: [String]
getDefaultLabels = ["TODO", "FIXME", "XXX"]

getDedaultExtensions :: [String]
getDedaultExtensions = ["php", "js", "xml", "yml", "yaml"]

colorGreen :: String -> String
colorGreen input = "\x1b[32m" ++ input ++ "\x1b[0m"

colorRed :: String -> String
colorRed input = "\x1b[31m" ++ input ++ "\x1b[0m"

textWithBackground :: String -> String -> String
textWithBackground color input = "\x1b[" ++ color ++ "m" ++ input ++ "\x1b[0m"

colorBlue :: String -> String
colorBlue input = "\x1b[34m" ++ input ++ "\x1b[0m"

textWithRedBackground :: String -> String
textWithRedBackground input = textWithBackground "41;1;37" (" " ++ input ++ " ")

textWithGreenBackground :: String -> String
textWithGreenBackground input = textWithBackground "42;1;37" (" " ++ input ++ " ")

spacesWithRedBackground :: Int -> String
spacesWithRedBackground count = textWithBackground "41;1;37" (replicate count ' ')

spacesWithGreenBackground :: Int -> String
spacesWithGreenBackground count = textWithBackground "42;1;37" (replicate count ' ')

removeBeforeLabel :: String -> String -> String
removeBeforeLabel input targetLabel = case dropWhile (not . isPrefixOf targetLabel) (tails input) of
    (rest:_) -> rest
    _ -> input

removeBeforeLabels :: String -> [String] -> String
removeBeforeLabels = foldl removeBeforeLabel